File: verifiers/javascript/hash_batch.js
const { computeBlake3, merkleLeafHash, computeMerkleRoot, toHex } = require('./verifier');

/**
 * Strictly decode a base64 string, exiting on invalid input.
 * Buffer.from(str, 'base64') silently ignores invalid characters in Node.js,
 * so we re-encode and compare to detect malformed input.
 */
function strictBase64Decode(b64, index) {
  const buf = Buffer.from(b64, 'base64');
  const roundTrip = buf.toString('base64');
  // Normalize: strip trailing '=' padding for comparison
  if (roundTrip.replace(/=+$/, '') !== b64.replace(/=+$/, '')) {
    process.stderr.write(`invalid base64 at index ${index}\n`);
    process.exit(1);
  }
  return buf;
}

let input = '';
process.stdin.setEncoding('utf8');
process.stdin.on('data', (chunk) => {
  input += chunk;
});

process.stdin.on('end', () => {
  const request = JSON.parse(input);
  // Default to "blake3" for backward compatibility
  const op = request.op || 'blake3';
  let hashes;

  switch (op) {
    case 'blake3':
      hashes = request.records_b64.map((recordB64, index) => {
        const data = strictBase64Decode(recordB64, index);
        return toHex(computeBlake3(data));
      });
      break;

    case 'merkle_leaf_hash':
      hashes = request.records_b64.map((recordB64, index) => {
        const data = strictBase64Decode(recordB64, index);
        return toHex(merkleLeafHash(data));
      });
      break;

    case 'merkle_root': {
      // All records_b64 are leaves of a single tree.
      // computeMerkleRoot already returns a hex string.
      const leaves = request.records_b64.map((recordB64, index) => {
        return new Uint8Array(strictBase64Decode(recordB64, index));
      });
      const root = computeMerkleRoot(leaves);
      hashes = [root];
      break;
    }

    default:
      process.stderr.write(`unknown op: ${op}\n`);
      process.exit(1);
  }

  process.stdout.write(JSON.stringify({ hashes }) + '\n');
});

File: verifiers/rust/src/bin/hash_batch.rs
use base64::Engine;
use olympus_verifier::{compute_blake3, compute_merkle_root, merkle_leaf_hash};
use serde::{Deserialize, Serialize};
use std::io::{self, Read};

#[derive(Deserialize)]
struct HashBatchRequest {
    #[serde(default)]
    op: Option<String>,
    records_b64: Vec<String>,
}

#[derive(Serialize)]
struct HashBatchResponse {
    hashes: Vec<String>,
}

fn main() {
    let mut input = String::new();
    if let Err(err) = io::stdin().read_to_string(&mut input) {
        eprintln!("failed to read stdin: {err}");
        std::process::exit(1);
    }

    let req: HashBatchRequest = match serde_json::from_str(&input) {
        Ok(req) => req,
        Err(err) => {
            eprintln!("failed to parse request JSON: {err}");
            std::process::exit(1);
        }
    };

    // Default to "blake3" for backward compatibility
    let op = req.op.as_deref().unwrap_or("blake3");

    let hashes = match op {
        "blake3" => {
            let mut result = Vec::with_capacity(req.records_b64.len());
            for (idx, record_b64) in req.records_b64.iter().enumerate() {
                let record =
                    match base64::engine::general_purpose::STANDARD.decode(record_b64.as_bytes()) {
                        Ok(bytes) => bytes,
                        Err(err) => {
                            eprintln!("invalid base64 at index {idx}: {err}");
                            std::process::exit(1);
                        }
                    };
                result.push(hex::encode(compute_blake3(&record)));
            }
            result
        }
        "merkle_leaf_hash" => {
            let mut result = Vec::with_capacity(req.records_b64.len());
            for (idx, record_b64) in req.records_b64.iter().enumerate() {
                let record =
                    match base64::engine::general_purpose::STANDARD.decode(record_b64.as_bytes()) {
                        Ok(bytes) => bytes,
                        Err(err) => {
                            eprintln!("invalid base64 at index {idx}: {err}");
                            std::process::exit(1);
                        }
                    };
                result.push(hex::encode(merkle_leaf_hash(&record)));
            }
            result
        }
        "merkle_root" => {
            // All records_b64 are leaves of a single tree
            let mut leaves = Vec::with_capacity(req.records_b64.len());
            for (idx, record_b64) in req.records_b64.iter().enumerate() {
                let record =
                    match base64::engine::general_purpose::STANDARD.decode(record_b64.as_bytes()) {
                        Ok(bytes) => bytes,
                        Err(err) => {
                            eprintln!("invalid base64 at index {idx}: {err}");
                            std::process::exit(1);
                        }
                    };
                leaves.push(record);
            }
            match compute_merkle_root(&leaves) {
                Ok(root) => vec![root],
                Err(err) => {
                    eprintln!("merkle_root failed: {err}");
                    std::process::exit(1);
                }
            }
        }
        _ => {
            eprintln!("unknown op: {op}");
            std::process::exit(1);
        }
    };

    let response = HashBatchResponse { hashes };
    match serde_json::to_string(&response) {
        Ok(json) => println!("{json}"),
        Err(err) => {
            eprintln!("failed to serialize response JSON: {err}");
            std::process::exit(1);
        }
    }
}
